server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java

```diff
@@ -132,9 +130,9 @@ public void removeAttribute(String key) {
 
     @Override
     public Iterator<String> getAttributeNames() {
-        Vector<String> names = new Vector<>();
-        for (Enumeration<String> e = attributes.keys(); e.hasMoreElements(); ) {
-            names.add(e.nextElement());
+        final List<String> names;
+        synchronized (attributes) {
```

Comment:
I don't like that **synchronize** at all.

I had a quick look, and **attributes** can be a **ConcurrentHashMap**.

This method become:

```
    @Override
    public Iterator<String> getAttributeNames() {
        return attributes.keySet().iterator();
    }
```

Sources: http://javarevisited.blogspot.com/2011/04/difference-between-concurrenthashmap.html#axzz50RtNZKVw

Reply:
I traced the introduction of `Hashtable attributes` all the way back to [revision 107158](https://svn.apache.org/viewvc?view=revision&revision=107158) of `james/server/trunk/src/org/apache/james/James.java` on September 13, 2000.

I'm not seeing why it needs to be a synchronized collection.

The only other implementation of `MailetContext` that I found is `FakeMailContext` that is used in test code. `FakeMailContext` uses a `HashMap`.

I think it's safe to use a `HashMap` in `JamesMailetContext` as well.